### DISCUSSION TITLE
Can you make a local version for steam please?

### DISCUSSION THREAD

**Discussion by 715883226** [Aktanusa/CookieMonster#1003] [Q&A]
From China, and the internet sucks, so when I use the code to load this mod, I can't load it every time, it makes me so depressed.
The function of this mod cannot be replaced. So I specially signed up this account to ask for help.

**Comment by 715883226** [comment]
(´•ω•`๑)

**Comment by DanielNoord** [comment]
I don't really know what you mean with "local mod" and how we should do that. I don't have access to the Steam version because I work on Mac so I can't really do this myself. Sorry!

**Reply by 715883226** [reply]
oh... I see...I'm sorry I didn't make myself clear and wasted your time.

The current situation is that I use the code
【javascript: (function () {
  Game.LoadMod('https://cookiemonsterteam.github.io/CookieMonster/dist/CookieMonster.js');
}());】
in the steam version, and it shows
【net::ERR_CONNECTION_RESET】,
it works a few times, but most of the time it doesn't work, because I can't connect to GitHub smoothly all the time.
Considering that I had never studied code so that I could not change it myself, I came here to ask for help.
I will paste two pieces of code below, one is an example, the other is automatic buying and selling. No matter what the result is, thank you for replying to me!

**Sample mod that comes with steam version.**

Game.registerMod("sample mod",{
	init:function(){
		//this function is called as soon as the mod is registered
		//declare hooks here
		
		//note: this mod does nothing but show a notification at the bottom of the screen once it's loaded
		Game.Notify(`Example mod loaded!`,'',[16,5]);
	},
	save:function(){
		//use this to store persistent data associated with your mod
	},
	load:function(str){
		//do stuff with the string data you saved previously
	},
});

/*

Breakdown of info.txt:
	"Name": "Sample Mod",
		the visibly displayed name of your mod
	"ID": "sample mod",
		the text id of your mod (alphanumeric characters and spaces only), used as its key when saving or loading data
		also used as identifiers in other mods' dependencies
		this id must be the same as the first parameter used in your Game.registerMod(id,hooks)
	"Author": "Orteil",
		your name here!
	"Description": "A bare-bones sample mod.",
		a description of your mod
	"ModVersion": 1,
		a number value for your mod's version
	"GameVersion": 2.031,
		the last version of Cookie Clicker this mod was confirmed to run on
	"Date": "13/08/2021",
		the date your mod was released or last updated
	"Dependencies": [],
		optional; an array of IDs of other mods that must be loaded before this one, ie. ["cool mod preloader","extra stuff"]
	"LanguagePacks": ["lang.js"],
		optional; an array of local files containing localization data (ie. changing game text, adding translations etc)
	"Disabled": 1,
		optional; if set to 1, this mod will be disabled by default
	"AllowSteamAchievs":1,
		optional; by default, mods (unless they only consist of language files) block the unlocking of Steam achievements while enabled; set this to 1 if this is a good honest mod that does not incredibly unbalance the game

*/

**IdleTrading mod**

if(IdleTrading === undefined) var IdleTrading = {};
IdleTrading.name = 'Idle Trading';
IdleTrading.version = '1.10';
IdleTrading.GameVersion = '2.042';

IdleTrading.launch = function(){
	IdleTrading.defaultConfig = function(){
		
		var conf = {
			goods: [],
			autoBuy: 1,
			autoSell: 1
		};
		
		for(var iG = 2; iG < Game.ObjectsN; iG++){
			conf.goods.push({
				active: true,
				buyThresh: -1,
				sellThresh: -1,
				minPrice: 99999,
				maxPrice:-99999
			});
		}
		
		return conf;
	}

	IdleTrading.init = function(){
		IdleTrading.isLoaded = 1;
		
		IdleTrading.restoreDefaultConfig(1);
		
		IdleTrading.ReplaceGameMenu();
		CCSE.MinigameReplacer(IdleTrading.ReplaceNativeMarket, "Bank");
		
		
		//***********************************
		//    Post-Load Hooks 
		//    To support other mods interfacing with this one
		//***********************************
		if(IdleTrading.postloadHooks) {
			for(var i = 0; i < IdleTrading.postloadHooks.length; ++i) {
				IdleTrading.postloadHooks[i]();
			}
		}
		
		if (Game.prefs.popups) Game.Popup(IdleTrading.name + ' loaded!');
		else Game.Notify(IdleTrading.name + ' loaded!', '', '', 1, 1);
	}


	//***********************************
	//    Menu Replacer
	//***********************************
	
	IdleTrading.ReplaceGameMenu = function(){
		Game.customOptionsMenu.push(function(){
			CCSE.AppendCollapsibleOptionsMenu(IdleTrading.name, IdleTrading.getMenuString());
		});
		
		Game.customStatsMenu.push(function(){
			CCSE.AppendStatsVersionNumber(IdleTrading.name, IdleTrading.version);
		});
	}

	IdleTrading.getMenuString = function(){
		if(Game.Objects["Bank"].minigameLoaded){
			let m = CCSE.MenuHelper;
			var M = Game.Objects['Bank'].minigame;
			
			var str = 	'<div class="listing">' + m.ActionButton("IdleTrading.restoreDefaultConfig(2); Game.UpdateMenu();", 'Restore Default') +  
						(typeof InsugarTrading == 'undefined' ? '' : m.ActionButton("IdleTrading.importInsugarTrading(); Game.UpdateMenu();", 'Import from Insugar Trading')) + 
						'</div>' + 
						'<div class="listing">' + m.ToggleButton(IdleTrading.config, 'autoBuy', 'IdleTrading_autoBuyButton', 'AutoBuy ON', 'AutoBuy OFF', "IdleTrading.Toggle") +
												  m.ToggleButton(IdleTrading.config, 'autoSell', 'IdleTrading_autoSellButton', 'AutoSell ON', 'AutoSell OFF', "IdleTrading.Toggle") + '</div>';
			
			str += m.Header('Goods');
			
			for(var iG = 0; iG < M.goodsById.length; iG++){
				var me = M.goodsById[iG];
				var conf = IdleTrading.config.goods[iG];
				
				str += '<div class="listing" style="text-align:left;"><div class="icon" style="pointer-events:none;display:inline-block;transform:scale(0.5);margin:-16px -18px -16px -14px;vertical-align:middle;background-position:' + (-me.icon[0] * 48) + 'px ' + (-me.icon[1] * 48) + 'px;"></div><span class="bankSymbol" style="width:30px;overflow:hidden;white-space:nowrap;">' + me.symbol + '</span>';
				str += '<label> Buy at:</label>' + m.InputBox('IdleTrading_buyThresh_' + iG, 65, conf.buyThresh, 'IdleTrading.UpdatePref(' + iG + ', this.value, 0)');
				str += '<label> Sell at:</label>' + m.InputBox('IdleTrading_sellThresh_' + iG, 65, conf.sellThresh, 'IdleTrading.UpdatePref(' + iG + ', this.value, 1)');
				str += '<label>Historical min: <b>$' + Beautify(conf.minPrice, 2) + '</b>; Historical max: <b>$' + Beautify(conf.maxPrice, 2) + '</b></label>';
				str += '</div>';
			}
			
			return str;
		}
		else{
			return '<div class="listing">Stock market minigame not loaded!</div>';
		}
	}


	//***********************************
	//    Configuration
	//***********************************
	
	IdleTrading.save = function(){
		return JSON.stringify(IdleTrading.config);
	}

	IdleTrading.load = function(str){
		var config = JSON.parse(str);
			
		for(var pref in config){
			if(pref == "goods"){
				for(var iG = 0; iG < config.goods.length; iG++){
					if(IdleTrading.config.goods[iG]){
						for(var pref2 in config.goods[iG]){
							IdleTrading.config.goods[iG][pref2] = config.goods[iG][pref2];
						}
					}
				}
			}
			else{
				IdleTrading.config[pref] = config[pref];
			}
		}
	}

	IdleTrading.restoreDefaultConfig = function(mode){
		IdleTrading.config = IdleTrading.defaultConfig();
		if(mode == 2) IdleTrading.save(IdleTrading.config);
	}
	
	IdleTrading.Toggle = function(prefName, button, on, off, invert){
		if(IdleTrading.config[prefName]){
			l(button).innerHTML = off;
			IdleTrading.config[prefName] = 0;
		}
		else{
			l(button).innerHTML = on;
			IdleTrading.config[prefName] = 1;
		}
		l(button).className = 'option' + ((IdleTrading.config[prefName] ^ invert) ? '' : ' off');
	}
	
	IdleTrading.ToggleGood = function(goodID, button, on, off, invert){
		if (IdleTrading.config.goods[goodID]){
			l(button).innerHTML = off;
			IdleTrading.config.goods[goodID].active = 0;
		}else{
			l(button).innerHTML = on;
			IdleTrading.config.goods[goodID].active = 1;
		}
		
		l(button).className = 'option' + ((IdleTrading.config.goods[goodID].active^invert) ? '' : ' off');
		IdleTrading.save(IdleTrading.config);
	}
	
	IdleTrading.UpdatePref = function(goodID, value, mode){
		var val = parseFloat(value);
		if(!isNaN(val)){
			if(mode == 0) IdleTrading.config.goods[goodID].buyThresh = val;
			if(mode == 1) IdleTrading.config.goods[goodID].sellThresh = val;
		}
		Game.UpdateMenu();
	}
	
	IdleTrading.importInsugarTrading = function(){
		var config = IdleTrading.config;
		var quant = InsugarTrading.settings.quantilesToDisplay;
		var bankLevel = Game.Objects["Bank"].level;
		var sellThresh = -1;
		var buyThresh = 2;
		
		for(var i = 0; i < quant.length; i++){
			if(quant[i] < buyThresh) buyThresh = quant[i];
			if(quant[i] > sellThresh) sellThresh = quant[i];
		}
		
		for(var iG = 0; iG < config.goods.length; iG++){
			config.goods[iG]['buyThresh'] = Math.round(100 * InsugarTrading.quantile(bankLevel, iG, buyThresh)) / 100;
			config.goods[iG]['sellThresh'] = Math.round(100 * InsugarTrading.quantile(bankLevel, iG, sellThresh)) / 100;
		}
	}
	
	
	//***********************************
	//    Functionality
	//***********************************
	
	IdleTrading.ReplaceNativeMarket = function() {
		if(!Game.customMinigame['Bank'].tick) Game.customMinigame['Bank'].tick = [];
		Game.customMinigame['Bank'].tick.push(IdleTrading.Logic);
	}
	
	IdleTrading.Logic = function(){
		var M = Game.Objects['Bank'].minigame;
		for(var iG = 0; iG < M.goodsById.length; iG++){
			var good = M.goodsById[iG];
			var conf = IdleTrading.config.goods[iG];
			var price = Math.round(100 * M.getGoodPrice(good)) / 100;
			
			if(IdleTrading.config.autoBuy && conf.buyThresh != -1){
				if(price <= conf.buyThresh) M.buyGood(iG, 10000);
			}
			if(IdleTrading.config.autoSell && conf.sellThresh != -1){
				if(price >= conf.sellThresh) M.sellGood(iG, 10000);
			}
			
			if(price < conf.minPrice) conf.minPrice = price;
			if(price > conf.maxPrice) conf.maxPrice = price;
		}
	}
	
	
	if(CCSE.ConfirmGameVersion(IdleTrading.name, IdleTrading.version, IdleTrading.GameVersion)) Game.registerMod(IdleTrading.name, IdleTrading); // IdleTrading.init();
}

if(!IdleTrading.isLoaded){
	if(CCSE && CCSE.isLoaded){
		IdleTrading.launch();
	}
	else{
		if(!CCSE) var CCSE = {};
		if(!CCSE.postLoadHooks) CCSE.postLoadHooks = [];
		CCSE.postLoadHooks.push(IdleTrading.launch);
	}
}

Thanks again!

**Comment by DanielNoord** [comment]
Can't you just paste the content of https://cookiemonsterteam.github.io/CookieMonster/dist/CookieMonster.js and put it into a local file and load that?

**Reply by 715883226** [reply]
I dragged the js file directly into the game, but there was no response.
I also tried to directly replace the available mod‘s js file with cookiemonster’s js file, but it didn't work.
I also tried to study the code format of mod that can be used to port cookiemonster’s js file, but I didn't know the rules of java at all, and it didn't work.
( ´_ゝ｀)

**Reply by DanielNoord** [reply]
> I also tried to directly replace the available mod‘s js file with cookiemonster’s js file, but it didn't work.

Does the steam version have a console which outputs errors? I might be able to help you if I can see the errors this methods produces!

**Comment by 715883226** [comment]
It shows like this_(:з)∠)_

![QQ浏览器截图20210929103959](https://user-images.githubusercontent.com/91250025/135194068-b87155fb-85d4-4cf4-814c-cd4966c5a6d8.png)

**Reply by 715883226** [reply]
And... I just found out that it can be unfolded.(´•ω•`๑)
![QQ浏览器截图20210929104643](https://user-images.githubusercontent.com/91250025/135194618-e4b2e29c-10d2-4af8-bbed-2743315e76fb.png)

**Reply by DanielNoord** [reply]
You might want to retry with the latest version of the mod? Other than that I can offer no support atm as I don't have access to the steam version due to being on Mac. You might want to ask in the discord/reddit if people know how to help you!

**Reply by 715883226** [reply]
oh! okay, thank you_(:з)∠)_

**Reply by sabata3** [reply]
> You might want to retry with the latest version of the mod? Other than that I can offer no support atm as I don't have access to the steam version due to being on Mac. You might want to ask in the discord/reddit if people know how to help you!

Here are all the "sample" files from the steam version. I believe they include hints on how to format mods.
[SteamMods.zip](https://github.com/CookieMonsterTeam/CookieMonster/files/7274544/SteamMods.zip)

I hope to see progress, as there are a number of mods I'd like to use and it seems nothing works.
Also, it looks like Orteil is intending to add Steam Workshop support to the game (unless it's a default structure enforced by steam).
